src/main/java/org/elasticsearch/index/query/QueryBuilderSerializer.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.query;
+
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.indices.query.IndicesQueriesRegistry;
+
+import java.io.IOException;
+
+/**
+ * Reads and writes a {@link QueryBuilder} when serialized over the wire based on its name which gets serialized
+ * before the the query itself and allows, when reading, to create the proper {@link QueryBuilder} subclass instance.
+ */
+public class QueryBuilderSerializer {
+
+    @Inject
+    static IndicesQueriesRegistry indicesQueriesRegistry;
```

Comment:
I totally agree....then the idea would be to take out of guice's hands the whole queries registry, breaking also what plugins have to do when they register custom queries? I am not sure how we could do that otherwise (I guess we would add a registerParser static method somewhere). Or the other way around, of we want to make this per node, not sure how much it makes sense, I don't know how we could use the queries registry from the query builders, given that they are not created through guice. Suggestions?